src/query/execute_statement.c

```diff
@@ -17193,3 +17061,94 @@ do_send_plan_trace_to_session (PARSER_CONTEXT * parser)
       free_and_init (plan_str);
     }
 }
+
+static int
+do_preliminary_checks (PARSER_CONTEXT * parser, PT_NODE * statement, PT_NODE * class_, PT_NODE * update,
+		       PT_NODE * values)
+{
+  int error = NO_ERROR;
+  int upd_has_uniques = 0;
+  bool has_default_values_list = false;
+
+  /* Check if server allows an insert. */
+  error = is_server_insert_allowed (parser, statement);
+  if (error != NO_ERROR || statement->info.insert.server_allowed != SERVER_INSERT_IS_ALLOWED)
+    {
+      return error;
+    }
+
+  /* Check non null attrs. */
+  if (values->info.node_list.list_type == PT_IS_DEFAULT_VALUE)
+    {
+      has_default_values_list = true;
+    }
+
+  error =
+    check_missing_non_null_attrs (parser, statement->info.insert.spec, statement->info.insert.attr_list,
+				  has_default_values_list);
+  if (error != NO_ERROR)
+    {
+      ASSERT_ERROR ();
+      return error;
+    }
+
+  /* Test if server UPDATE is allowed */
+
+  if (statement->info.insert.odku_assignments != NULL)
+    {
+      update = do_create_odku_stmt (parser, statement);
+      if (update == NULL)
+	{
+	  error = ER_FAILED;
+	  return error;
+	}
+      if (statement->info.insert.server_allowed == SERVER_INSERT_IS_ALLOWED)
+	{
+	  int server_allowed = 0;
+	  error =
+	    is_server_update_allowed (parser, &statement->info.insert.odku_non_null_attrs, &upd_has_uniques,
+				      &server_allowed, update);
+
+	  if (error != NO_ERROR)
+	    {
+	      ASSERT_ERROR ();
+	      return error;
+	    }
+	  if (!server_allowed)
+	    {
+	      statement->info.insert.server_allowed = SERVER_INSERT_IS_NOT_ALLOWED;
+	      return error;
+	    }
+	}
+    }
+
+  /* fetch the class for instance write purpose */
+  if (!locator_fetch_class (class_->info.name.db_object, DB_FETCH_CLREAD_INSTWRITE))
```

Comment:
@lucianlazar2211 

`locator_fetch_class` call will set IX_LOCK on the table. Please examine this is the best place to hold the lock.
I think it would be enough to do it at the end of the function.